commands/http/handler.go

```diff
@@ -121,6 +121,16 @@ func (i internalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	i.ctx.Context = ctx
 	req.SetContext(i.ctx)
 
+	if cn, ok := w.(http.CloseNotifier); ok {
+		go func() {
+			select {
+			case <-cn.CloseNotify():
+				cancel()
+			case <-ctx.Done():
+			}
+		}()
+	}
+
```

Comment:
im starting to think it's got something to do with [the weird `i.ctx.Context = ctx` line](https://github.com/ipfs/go-ipfs/pull/1388/files#diff-4df02c629d914567df45314a464367b9R121) that @cryptix and I are both creeped out by. but i tried commenting it out and running the tests but they still fail all the same.